src/Sylius/Bundle/ShopBundle/Grid/ProductGrid.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ShopBundle\Grid;
+
+use Sylius\Bundle\GridBundle\Builder\Field\DateTimeField;
+use Sylius\Bundle\GridBundle\Builder\Field\Field;
+use Sylius\Bundle\GridBundle\Builder\Field\StringField;
+use Sylius\Bundle\GridBundle\Builder\Filter\StringFilter;
+use Sylius\Bundle\GridBundle\Builder\GridBuilderInterface;
+use Sylius\Bundle\GridBundle\Grid\AbstractGrid;
+use Sylius\Bundle\GridBundle\Grid\ResourceAwareGridInterface;
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Sylius\Component\Core\Model\TaxonInterface;
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Sylius\Component\Taxonomy\Repository\TaxonRepositoryInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Webmozart\Assert\Assert;
+
+final class ProductGrid extends AbstractGrid implements ResourceAwareGridInterface
+{
+    private ?Request $resquest;
+
+    /**
+     * @param TaxonRepositoryInterface<TaxonInterface> $taxonRepository
+     */
+    public function __construct(
+        private string $resourceClass,
+        private ChannelContextInterface $channelContext,
+        private TaxonRepositoryInterface $taxonRepository,
+        private LocaleContextInterface $localeContext,
+        RequestStack $requestStatck,
+        private bool  $includeAllDescendants,
+    ) {
+        $this->resquest = $requestStatck->getMainRequest();
+    }
+
+    public static function getName(): string
+    {
+        return 'sylius_shop_product';
+    }
+
+    public function buildGrid(GridBuilderInterface $gridBuilder): void
+    {
+        Assert::notNull($this->resquest);
+        $localeCode = $this->localeContext->getLocaleCode();
+
+        // @see Sylius\Bundle\ResourceBundle\ExpressionLanguage\NotNullExpressionFunctionProvider
+        $taxon = $this->taxonRepository->findOneBySlug(
+                    $this->resquest->attributes->get('slug'),
+                    $localeCode,
+        );
+        if ($taxon === null)  {
+            throw new NotFoundHttpException('Requested page is invalid');
+        }
```

Comment:
We might want to move that to a helper or somewhere?

Reply:
@diimpp Any ideas or thoughts on this one? I guess it would make sense to make a class or a trait that could emulate what we had in the expression parser.